### PR TITLE
src/runtime: switch to stronger hash function

### LINKED ISSUE #1553: maps need growing logic when they fill up
Tried to compile and run [this code](https://github.com/kokizzu/kokizzu-benchmark/blob/master/assoc-small/map.go), but it doesn't complete within 1000s (while other languages/compiler completed within 1-10s)

```
// map.go
package main

import (
        "bytes"
        "fmt"
        "strconv"
)

type MSD map[string]float64

const MAX_DATA = 1230000

var i2ch []byte

func init() {
        i2ch = []byte{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'B', 'c', 'D', 'e', 'F'}
}
func get_first_digit(d float64) int {
        for d > 10 {
                d /= 10
        }
        return int(d)
}
func to_rhex(v int) string {
        hex := bytes.Buffer{}
        for v > 0 {
                hex.WriteByte(i2ch[v%16])
                v /= 16
        }
        return hex.String()
}
func set_or_inc(m map[string]float64, key string, set, inc int, ctr *int) {
        if _, ok := m[key]; !ok {
                m[key] = float64(set)
        } else {
                m[key] += float64(inc)
                *ctr += 1
        }
}
func main() {
        m := MSD{}
        dup1, dup2, dup3 := 0, 0, 0
        for z := MAX_DATA; z > 0; z-- {
                val2 := MAX_DATA - z
                val3 := MAX_DATA*2 - z
                key1 := strconv.Itoa(z)
                key2 := strconv.Itoa(val2)
                key3 := to_rhex(val3)
                set_or_inc(m, key1, z, val2, &dup1)
                set_or_inc(m, key2, val2, val3, &dup2)
                set_or_inc(m, key3, val3, z, &dup3)
        }
        fmt.Println(dup1, dup2, dup3)
        total, verify, count := 0, 0, 0
        for k, v := range m {
                total += get_first_digit(v)
                verify += len(k)
                count += 1
        }
        fmt.Println(total, verify, count)
}
```
image

### PR DESCRIPTION
**Pull request by dgryski**
Fixes #1553

### CODE REVIEW

**Comment by dgryski** [comment]
This still needs build tags to keep the older fnv hash function on smaller platforms.

**Comment by dgryski** [comment]
```
~/go/src/github.com/tinygo-org/tinygo/testdata $ tinygo build -target=wasi -size=full mapgrowth.go
   code  rodata    data     bss |   flash     ram | package
------------------------------- | --------------- | -------
   6063       0     309  130588 |    6372  130897 | (unknown)
   1122       0       0       4 |    1122       4 | internal/task
   1654       0       0       0 |    1654       0 | main
   8556       0      16     155 |    8572     171 | runtime
------------------------------- | --------------- | -------
  17395       0     325  130747 |   17720  131072 | total
~/go/src/github.com/tinygo-org/tinygo/testdata $ tinygo build -tags=runtime_memhash_tsip -target=wasi -size=full mapgrowth.go
   code  rodata    data     bss |   flash     ram | package
------------------------------- | --------------- | -------
   6314       0     309  130588 |    6623  130897 | (unknown)
      5       0       0       0 |       5       0 | encoding/binary
   1122       0       0       4 |    1122       4 | internal/task
   1654       0       0       0 |    1654       0 | main
     24       0       0       0 |      24       0 | math/bits
   9194       0      16     155 |    9210     171 | runtime
------------------------------- | --------------- | -------
  18313       0     325  130747 |   18638  131072 | total
```

**Comment by dgryski** [comment]
Hmm... seems like the speed crossover for tsip vs fnv (at least on amd64 and wasi) is at ~20-30 bytes.  LevelDB seems to always be faster.  And smaller than tsip.

**Comment by dgryski** [comment]
```
~/go/src/github.com/tinygo-org/tinygo/testdata $ tinygo build -tags=runtime_memhash_leveldb -target=wasi -size=full mapgrowth.go
   code  rodata    data     bss |   flash     ram | package
------------------------------- | --------------- | -------
   6064       0     309  130588 |    6373  130897 | (unknown)
   1122       0       0       4 |    1122       4 | internal/task
   1654       0       0       0 |    1654       0 | main
   8663       0      16     155 |    8679     171 | runtime
------------------------------- | --------------- | -------
  17503       0     325  130747 |   17828  131072 | total
```

**Comment by deadprogram** [comment]
@dgryski can you please rebase this PR now against `dev`? Thank you.

**Comment by dgryski** [comment]
Rebased against the hash seed PR.  Will rebase against dev when it's merged.

**Comment by deadprogram** [comment]
@dgryski I think this is ready for rebase.

**Comment by dgryski** [comment]
Rebased.

File: src/runtime/memhash_tsip.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	s.round()
+	s.v0 ^= b
+
+	// finalization
+	s.v1 ^= 0xff
+	s.round()
+	s.v1 = bits.RotateLeft32(s.v1, 16)
+	s.round()
+	s.v1 = bits.RotateLeft32(s.v1, 16)
+	//s.round()
```
Comment (deadprogram) [inline src/runtime/memhash_tsip.go]: Is this supposed to be commented here?

**Comment by deadprogram** [comment]
Other than the one commented line, seems to me to be ready to go.

File: src/runtime/memhash_tsip.go (hunk 1 above)
Comment (dgryski) [reply src/runtime/memhash_tsip.go]: I'll remove that. Left over from when I was designing the hash function.

**Comment by dgryski** [comment]
Rebased and commented line removed.

**Comment by deadprogram** [comment]
Last call for feedback before I merge!

File: src/runtime/memhash_tsip.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+	pslice := (*_bslice)(unsafe.Pointer(&p))
+	pslice.ptr = (*byte)(ptr)
+	pslice.cap = n
+	pslice.len = n
+
+	return p
+}
+
+// tsip hash -- github.co/dgryski/tsip
```
Comment (aykevl) [inline src/runtime/memhash_tsip.go]: Typo!

**Review by aykevl — COMMENTED** [review]
This looks reasonable.
It would be nice though to have a short description in the memhash files where the algorithm comes from and their strength/performance. Just 1-2 sentences to get an idea.

File: src/runtime/memhash_tsip.go (hunk 2 above)
Comment (dgryski) [reply src/runtime/memhash_tsip.go]: Nice catch.  I'll fix this and add some comments about the hash functions.

**Comment by dgryski** [comment]
Sample benchmark code.  Not sure where to put this..

```
package main

import (
        "hash/maphash"
        "strconv"
        "testing"
)

var buf [8192]byte

func BenchmarkMaphash(b *testing.B) {
        var h maphash.Hash
        benchmarkHash(b, "maphash", h)
}

func benchmarkHash(b *testing.B, str string, h maphash.Hash) {
        var sizes = []int{1, 2, 3, 4, 5, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 1024, 8192}
        for _, n := range sizes {
                b.Run(strconv.Itoa(n), func(b *testing.B) { benchmarkHashn(b, int64(n), h) })
        }
}

var total uint64

func benchmarkHashn(b *testing.B, size int64, h maphash.Hash) {
        b.SetBytes(size)

        sum := make([]byte, 4)

        for i := 0; i < b.N; i++ {
                h.Reset()
                h.Write(buf[:size])
                sum = h.Sum(sum[:0])
                total += uint64(sum[0])
        }
}
```

**Comment by deadprogram** [comment]
@aykevl what do you think?

**Comment by deadprogram** [comment]
I suggest that the decision about that benchmark code be put into a separate PR, so we can get this merged now.

Thank you for all your great work on this @dgryski

**Pull request merged**